vllm/attention/backends/flash_attn.py

```diff
@@ -207,32 +234,32 @@ def forward(
                 )
             else:
                 # prefix-enabled attention
-                output = PagedAttention.forward_prefix(
-                    query,
-                    key,
-                    value,
-                    key_cache,
-                    value_cache,
-                    attn_metadata.block_tables,
-                    attn_metadata.subquery_start_loc,
-                    attn_metadata.prompt_lens_tensor,
-                    attn_metadata.context_lens,
-                    attn_metadata.max_subquery_len,
-                    self.alibi_slopes,
+                output = flash_attn_varlen_func(
+                    q=query,
+                    k=key_cache,
+                    v=value_cache,
+                    cu_seqlens_q=attn_metadata.seq_start_loc,
+                    cu_seqlens_k=attn_metadata.context_lens,  # FIXME
+                    max_seqlen_q=attn_metadata.max_prompt_len,
+                    max_seqlen_k=attn_metadata.max_context_len,
+                    softmax_scale=self.scale,
+                    causal=True,
+                    window_size=self.sliding_window,
+                    alibi_slopes=self.alibi_slopes,
+                    block_table=attn_metadata.block_tables,
                 )
         else:
             # Decoding run.
-            output = PagedAttention.forward_decode(
-                query,
+            # TODO(skrider): tune num_splits heuristic
+            output = flash_attn_with_kvcache(
+                query.unsqueeze(1),
                 key_cache,
                 value_cache,
-                attn_metadata.block_tables,
-                attn_metadata.context_lens,
-                attn_metadata.max_context_len,
-                attn_metadata.kv_cache_dtype,
-                self.num_kv_heads,
-                self.scale,
-                self.alibi_slopes,
+                block_table=attn_metadata.block_tables,
+                cache_seqlens=attn_metadata.context_lens,
+                softmax_scale=self.scale,
+                causal=True,
+                alibi_slopes=self.alibi_slopes,
```

Comment:
should we allow to specify num_splits using env var? I think it is automatically chosen, but wonder if we should expose this

Reply:
As you might know, the `flash-attn` library includes heuristics to choose `num_splits`. Since this value highly depends on the input shapes, it does not make sense to set this as an env variable.